File: 474.onesAndZeroes/onesAndZeroes.cpp
// clang++ onesAndZeroes.cpp -std=c++11

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int findMaxForm(vector<string>& strs, int m, int n) {
        vector<vector<int> > dp(m + 1, vector<int>(n + 1, 0));
        for (string& str : strs) {
            int num_of_zeroes = 0, num_of_ones = 0;
            for (char c : str) {
                if (c == '0')
                    num_of_zeroes++;
                else if (c == '1')
                    num_of_ones++;
            }
            for (int i = m; i >= num_of_zeroes; i--) {
                for (int j = n; j >= num_of_ones; j--) {
                    dp[i][j] = max (dp[i][j], dp[i - num_of_zeroes][j - num_of_ones] + 1);
                }
            }
        }
        return dp[m][n];
    }
};

int main()
{
    Solution s;
    vector<string> arr = {"10", "0001", "111001", "1", "0"};
    cout << s.findMaxForm(arr, 5, 3) << endl;
    return 0;
}

File: 474.onesAndZeroes/onesAndZeroes.java
class Solution {
    public int findMaxForm(String[] strs, int m, int n) {
        int[][][] dp = new int[strs.length + 1][m + 1][n + 1];
        for (int k = 1; k <= strs.length; k++) {
            int[] nums = getNums(strs[k - 1]);
            for (int i = 0; i <= m; i++) {
                for (int j = 0; j <= n; j++) {
                    if (i >= nums[0] && j >= nums[1]) {
                        dp[k][i][j] = Math.max(dp[k - 1][i][j], dp[k - 1][i - nums[0]][j - nums[1]] + 1);
                    } else {
                        dp[k][i][j] = dp[k - 1][i][j];
                    }
                }
            }
        }
        return dp[strs.length][m][n];
    }
    
    public int[] getNums(String str) {
        int[] nums = new int[2];
        for (char c : str.toCharArray()) {
            if (c == '0') nums[0]++;
            else if (c == '1') nums[1]++;
        }
        return nums;
    }
}
